### PR TITLE
JAMES-3122 Upgrade to log4J2 in server/app

### PR DESCRIPTION
**Pull request by chibenwa**
LOG4J 1 is deprecated and had been reported not to play well with Java 9+

https://www.mail-archive.com/server-user@james.apache.org/msg16315.html

### CODE REVIEW

File: server/app/pom.xml [hunk 1]
```diff
@@ -548,6 +548,10 @@
             <artifactId>xbean-spring</artifactId>
             <scope>runtime</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-slf4j-impl</artifactId>
```
Comment (mbaechler) [inline server/app/pom.xml]: why not in runtime scope?

File: server/container/spring/pom.xml [hunk 2]
```diff
@@ -152,6 +147,15 @@
             <groupId>org.apache.commons</groupId>
             <artifactId>commons-configuration2</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j</artifactId>
+            <version>2.13.1</version>
```
Comment (mbaechler) [inline server/container/spring/pom.xml]: duplicate version?

**Comment by Arsnael** [comment]
```
[a8be09f2ee1739f538e21911c14afa4d606ac919] [ERROR] Failed to execute goal on project james-server-spring: Could not resolve dependencies for project org.apache.james:james-server-spring:jar:3.5.0-SNAPSHOT: Could not find artifact org.apache.logging.log4j:log4j:jar:2.13.1 in central (https://repo.maven.apache.org/maven2), try downloading from https://logging.apache.org/log4j/2.x/download.html -> [Help 1]
```
test this please

**Comment by chibenwa** [comment]
```
[be58e6b54c7b6752d461282b101dddaf4b86a6a3] [INFO] ------------------------------------------------------------------------
[be58e6b54c7b6752d461282b101dddaf4b86a6a3] [ERROR] Failed to execute goal on project james-server-spring: Could not resolve dependencies for project org.apache.james:james-server-spring:jar:3.5.0-SNAPSHOT: Could not find artifact org.apache.logging.log4j:log4j:jar:2.13.1 in central (https://repo.maven.apache.org/maven2), try downloading from https://logging.apache.org/log4j/2.x/download.html -> [Help 1]
```

https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j/2.13.1 ???

**Comment by chibenwa** [comment]
```
[be58e6b54c7b6752d461282b101dddaf4b86a6a3] [INFO] ------------------------------------------------------------------------
[be58e6b54c7b6752d461282b101dddaf4b86a6a3] [ERROR] Failed to execute goal on project james-server-spring: Could not resolve dependencies for project org.apache.james:james-server-spring:jar:3.5.0-SNAPSHOT: Could not find artifact org.apache.logging.log4j:log4j:jar:2.13.1 in central (https://repo.maven.apache.org/maven2), try downloading from https://logging.apache.org/log4j/2.x/download.html -> [Help 1]
```

https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j/2.13.1 ???

**Comment by chibenwa** [comment]
Pushed to solve conflict

**Comment by chibenwa** [comment]
Merged

**Comment by tommyhp2** [comment]
From the user mailing list, Rene Cordier asked me to look at this PR.  From what I've seen of the logging used in James 3.4.0, I think there are too many mechanisms:

commons-logging
slf4j
log4j (v1)
logback

Why not just use log4j2 primarily?  This what I've done:

\# ls -al *log* *slf*
-rw-r--r--  1 root  wheel   61829 Dec  3  2016 commons-logging-1.2.jar
-rw-r--r--  1 root  wheel   60201 Jul  5  2019 jboss-logging-3.4.0.Final.jar
lrwxr-xr-x  1 root  wheel      31 Mar 19 04:17 log4j-1.2-api-2.13.1.jar -> ../new/log4j-1.2-api-2.13.1.jar
lrwxr-xr-x  1 root  wheel      27 Mar 19 04:17 log4j-api-2.13.1.jar -> ../new/log4j-api-2.13.1.jar
lrwxr-xr-x  1 root  wheel      28 Mar 19 04:17 log4j-core-2.13.1.jar -> ../new/log4j-core-2.13.1.jar
lrwxr-xr-x  1 root  wheel      27 Mar 19 04:17 log4j-jcl-2.13.1.jar -> ../new/log4j-jcl-2.13.1.jar
lrwxr-xr-x  1 root  wheel      27 Mar 19 04:17 log4j-jul-2.13.1.jar -> ../new/log4j-jul-2.13.1.jar
lrwxr-xr-x  1 root  wheel      34 Mar 19 04:17 log4j-slf4j-impl-2.13.1.jar -> ../new/log4j-slf4j-impl-2.13.1.jar
lrwxr-xr-x  1 root  wheel      34 Mar 19 04:17 log4j-slf4j-impl-2.13.1.jar -> ../new/log4j-slf4j-impl-2.13.1.jar
-rw-r--r--  1 root  wheel  471901 Jul  5  2017 logback-core-1.2.3.jar
-rw-r--r--  1 root  wheel   10439 Sep  5  2019 metrics-logger-3.4.0.jar
lrwxr-xr-x  1 root  wheel      27 Mar 19 04:17 slf4j-api-1.7.30.jar -> ../new/slf4j-api-1.7.30.jar

and my log4j2 configuration replacement:

[https://github.com/tommyhp2/ajs/blob/master/src/main/webapp/WEB-INF/classes/log4j2.xml](url)

I've added the log4j2-jul bridge for extra coverage to ensure that I would see all logging.  If you intend to use slf4j as the primary logging, I believe you'll end up with these jars:

commons-logging
jcl-over-slf4j
log4j-over-slf4j
slf4j-api
log4j-core
log4j-api
log4j-to-slf4j

Even though I have not used slf4j (either in code or in implementation), IMO, I think log4j2 is better since the user can send the logs to any where (JDBC, email, etc) as needed.

Also, as of the official 3.4 distribution, there's only one logging configuration (log4j.properties) while there are many mechanisms used :(  My log4j2 replacement hack allow James 3.4 to run on JDK 11 and allow me to see more of the logging.

**Comment by chibenwa** [comment]
Hello @tommyhp2 ,

Guice products are relying on logback, while spring relies on log4j.

Inside the application we rely on slf4j to produce logs that will then be written either with log4j or slf4j (logging facade). Commons-logging is used by some dependency and directly bridged to slf4j.

> and my log4j2 configuration replacement

We need a pull request for that. Can you take care of it?

> Even though I have not used slf4j (either in code or in implementation), IMO, I think log4j2 is better since the user can send the logs to any where (JDBC, email, etc) as needed.

Logback can do all of that too.

Regarding slf4j vs log4j, I would stand for responsibility segragation:
 - SLF4J for producing logs (and it's not easy to change as we use it everywhere)
 - log4j2 OR logback for writing the logs.

Does it makes sense?

**Comment by tommyhp2** [comment]
@chibenwa ,

Yes, thank you for the clarification regarding logging facilities.  But which is the main?  I only see one configuration files for the log4j (v1).  Shouldn't there be at least another configuration file especially  for SLF4j since there bridges for it?  I don't know how well SLF4j and logback handles the bridging but when I was experimenting with log4j2 a while back, it handles all the dependencies well with the proper bridge jars in place.

Regarding the pull request, I'm still new to this github as software development is not my main until recently.  I've made the changes per documentations.  If it doesn't work, please let me know and I'll check it again.

**Comment by chibenwa** [comment]
I did open a pull request to upgrade the default configuration file for Spring logging https://github.com/linagora/james-project/pull/3234

**Pull request closed without merging**